Abstracting away the gpu

parthraut opened the pull request:
PR is mostly done, few notes:

1. ensure_homogeneous has not been implemented, I was a little confused on what it should check. Should it check that all gpus are the same kind? Ex. all GPUs must be the same kind and GPU architecture
2. Some methods for AMD GPU class have not been implemented, as well as handling the exceptions. I focused on NVIDIA gpus for this PR, but extending this should not be difficult if needed. 
3. Style: Failing because many names have uppercase letters: "N802 Function name `getTotalEnergyConsumption` should be lowercase". We decided on camel case to match pynvml methods, so what should be done to resolve this?

parthraut committed:
basic plan

parthraut committed:
stuff

parthraut committed:
added some functions in GPU classes

parthraut committed:
changed src in zeus master

parthraut committed:
made changes, need to test

parthraut committed:
trying to fix errors

parthraut committed:
failing test_power_limit_optimizer

parthraut committed:
replaced all usages, failing a lot :(

parthraut committed:
fixed testing issues

parthraut committed:
added some exception handling

parthraut committed:
cleaned up gpu class

parthraut committed:
added style, replaced pynvml exceptions

parthraut committed:
fixed some style

parthraut committed:
removed unnecessary file

jaywonchung commented:
Could you rebase to master? There was a small change in docs/requirements.txt.

jaywonchung commented:
Thanks a lot for your work! I'll look at the code soon.

Responses to your questions:
1. I think ensuring that the name of the devices (`nvmlDeviceGetName`) are equal is enough.
2. Sounds good!
3. See bottom of pyproject.toml. There you can disable specific Ruff error code on entire files.

jaywonchung commented on zeus/device/__init__.py:
Module-level docstring are missing. Please take a look at other modules' `__init__.py` and create something similar. You'll also need to add `device` to `zeus/__init__.py`.

jaywonchung commented on zeus/device/__init__.py:
```diff
@@ -0,0 +1 @@
+from zeus.device.gpu import get_gpus, ZeusNotSupportedGPUError, ZeusNoPermissionGPUError
```
I don't think it's good to selectively import some errors and not others.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+        try:
+            self.handle = pynvml.nvmlDeviceGetHandleByIndex(self.gpu_index)
+        except pynvml.NVMLError as e:
+            exception_class = NVIDIAGPU._exception_map.get(e.value, ZeusBaseGPUError)
```
Isn't this handled by `_handle_nvml_errors`?

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+        resetGpuLockedClocks(): Resets the GPU locked clocks to default values using `pynvml.nvmlDeviceResetGpuLockedClocks`.
+        getPowerUsage(): Returns the current power usage of the GPU using `pynvml.nvmlDeviceGetPowerUsage`.
+        supportsGetTotalEnergyConsumption(): Checks if the GPU supports retrieving total energy consumption, based on GPU architecture using `pynvml.nvmlDeviceGetArchitecture` and comparing with `pynvml.NVML_DEVICE_ARCH_VOLTA`.
+    getTotalEnergyConsumption(): Returns the total energy consumption of the GPU since the last driver load using `pynvml.nvmlDeviceGetTotalEnergyConsumption`.
```
Remove. The list of methods is auto-generated with mkdocs in the source code reference like [this](https://zeus-git-23-abstracting-away-the-gpu-jaywonchungs-projects.vercel.app/reference/device/gpu/#zeus.device.gpu.NVIDIAGPU).

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+    errors is integrated, mapping NVML error codes to custom exceptions for clearer error reporting and management.
+
+    Attributes:
+        _exception_map (dict): A dictionary mapping NVML error codes to custom exception classes for this application.
```
Remove. Only public attributes (i.e., attributes that are meant to be part of a stable public API exposed by the class) are to be noted in the class docstring.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+"""GPU management module for Zeus."""
```
This docstring should include the following information at the minimum:
- Class hierarchy
- The abstraction provided by this module (most importantly, `get_gpus`). Basically, how a Zeus-developer should use this module.
- How different GPU vendors are handled

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+
+
+class ZeusInitGPUError(ZeusBaseGPUError):
+    """Zeus GPU exception class Wrapper for ImportError or Failed to Initialize GPU libraries."""
```
```suggestion
    """Import error or GPU library initialization failures."""
```
Let's keep it concise.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+
+
+class GPU(abc.ABC):
+    """Abstract base class for GPU management. This class defines the interface for interacting with GPUs, including power management, clock settings, and information retrieval. Subclasses should implement the methods to interact with specific GPU libraries (e.g., NVML for NVIDIA GPUs)."""
```
The first line of a docstring should be a single sentence. Then, the next paragraph goes into more detail. See [Google Python style guide](https://android.googlesource.com/platform/external/google-styleguide/+/refs/tags/android-s-beta-2/pyguide.md#3_8_4-classes).

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+
+    @abc.abstractmethod
+    def _get_handle(self):
+        """Acquire and set the handle for the GPU, enabling further operations."""
```
A private abstract method doesn't make sense. You're designating something as an abstract method either because (1) it is a public API exposed as part of the type's abstraction (which is not the case since this starts with an underscore, meaning it's not public), or (2) some other existing non-abstract method in the base class relies on being able to call that method (which does not happen in `GPU`). Indeed, even if you just remove `_get_handle` from `GPU`, you'll notice that things will still work well. You can even move the content of `NVIDIAGPU._get_handle` into `NVIDIAGPU.__init__` and it'll be fine.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+        pynvml.NVML_ERROR_NOT_FOUND: ZeusNotFoundGPUError,
+        pynvml.NVML_ERROR_INSUFFICIENT_SIZE: ZeusInsufficientSizeGPUError,
+        pynvml.NVML_ERROR_INSUFFICIENT_POWER: ZeusInsufficientPowerGPUError,
+        pynvml.NVML_ERROR_DRIVER_NOT_LOADED: ZeusDriverErrorGPUError,
```
```suggestion
        pynvml.NVML_ERROR_DRIVER_NOT_LOADED: ZeusGPUDriverNotLoadedError,
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+""" EXCEPTION WRAPPERS """
+
+
+class ZeusInitGPUError(ZeusBaseGPUError):
```
Names are all kinda weird? I think it should all be named `ZeusGPUInitError` and such. When people look at the name of the error, they'll segment it as `Zeus InitGPU Error`. It might be fine for this one, but errors like `Zeus InsufficientPowerGPU Error` could even be misleading and `ZeusLostGPUGPUError` just looks strange.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+                return func(*args, **kwargs)
+            except pynvml.NVMLError as e:
+                exception_class = NVIDIAGPU._exception_map.get(
+                    e.value, ZeusBaseGPUError
```
I think we should fall back to `ZeusGPUUnknownError`, not the base error.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+    def setPersistenceMode(self) -> None:
+        """Enables persistence mode for the specified GPU."""
+        # TODO(JW): Check SYS_ADMIN permissions and error with an explanation.
+        pynvml.nvmlDeviceSetPersistenceMode(self.handle, pynvml.NVML_FEATURE_ENABLED)
```
This removed the ability to turn off persistence mode. Also, can you return the `TODO` comment to where this method originally was?

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,882 @@
+        pynvml.nvmlDeviceSetPersistenceMode(self.handle, pynvml.NVML_FEATURE_ENABLED)
+
+    @_handle_nvml_errors
+    def setPowerManagementLimit(self, value: int = None) -> None:
```
I detected some problems:
- Wrong type annotation. It should be `int | None`.
- More fundamentally, this breaks API consistency. The convention of returning a GPU configuration knob to its default value is called `reset` throughout NVML at least (e.g., `resetGpuLockedClocks`). It would actually be fine if you were adding `def resetPowerManagementLimit(self) -> None`.

parthraut committed:
fixed style, added support for ensure_homogeneous

parthraut committed:
fixed small issues

parthraut committed:
github giving weird error, not showing locally.

parthraut commented:
- added implementation for ensure_homogeneous
- modified pyproject.toml to fix style, but it still looks like the tests are failing on github actions with some weird error. I don't see this error when I run locally, it all passes.
- The only file left to be finished is PerseusOptimizer in zeus/optimizer/perseus/optimizer.py. In the perseus Optimizer class:

class PerseusOptimizer(Callback):
    """Perseus optimizer."""

    def __init__(
        self,
        rank: int,
        dp_rank: int,
        pp_rank: int,
        tp_rank: int,
        device_id: int,
        dp_degree: int,
        pp_degree: int,
        tp_degree: int,
        world_size: int,
        server_url: str,
        job_metadata: str | None = None,
    ) -> None:
        """Initialize the Perseus optimizer.

        Assumptions:
            - `torch.distributed` has been initialized.
            - `torch.cuda.set_device` has been called with `device_id`.
                This is needed to broadcast the job ID to all ranks.

        The master process (rank 0) will register the job with the Peresus
        server and retrieve the job ID of this job. Then, each rank will
        report itself to the Perseus server with the job ID.

        Args:
            rank: Global rank of the current process.
            dp_rank: Rank in the data parallel group.
            pp_rank: Rank in the pipeline parallel group.
            tp_rank: Rank in the tensor parallel group.
            **device_id: CUDA device ID that the current process manages.**
            dp_degree: Size of the data parallel group.
            pp_degree: Size of the pipeline parallel group.
            tp_degree: Size of the tensor parallel group.
            world_size: Total number of ranks that participate in training.
            server_url: URL of the Perseus server.
            job_metadata: An optional arbitrary string that describes the job. This will
                be appended to the job ID if given. Typically for logging purposes.
        """
        
Here, is device_id reindexed to respect CUDA_VISIBLE_DEVICES?

for example, if there are four GPUs on system and CUDA_VISIBLE_DEVICES="0,2", to access GPU with CUDA_VISIBLE_DEVICE "2", does it reindex it to be 1 or keep it as 2? Because ZeusMonitor reindexes, but from the comment above I wasn't sure

^ after doing some research, I assume its reindexed. Finished optimizer.py with this assumption.

parthraut committed:
finished optimizer code, fixed style

jaywonchung commented:
Correct, `1` is the index that will map to NVML device `2`. The test is failing because you imported `torch` in `gpu.py`. We actually avoid that in Zeus because we want people to be able to use Zeus without having PyTorch installed -- Someone might be using JAX, someone might just want to measure GPU energy and they're not even doing deep learning, etc. That's why we do this:
https://github.com/ml-energy/zeus/blob/f9275786fd98920b3bf98485c78f9e15b2304364/zeus/util/framework.py#L29-L42

jaywonchung commented on tests/test_monitor.py:
```diff
@@ -37,19 +38,24 @@
+@pytest.fixture(autouse=True, scope="function")
+def reset_gpus() -> None:
+    print("Resetting _gpus to None")
+    zeus.device.gpu._gpus = None
```
```suggestion
@pytest.fixture(autouse=True, scope="function")
def reset_gpus() -> None:
    """Reset the global variable `_gpus` to None on every test."""
    zeus.device.gpu._gpus = None
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+
+    @abc.abstractmethod
+    def getTotalEnergyConsumption(self) -> int:
+        """Return the total energy consumption of the GPU."""
```
```suggestion
        """Return the total energy consumption of the GPU since driver load."""
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+    errors is integrated, mapping amdsmi error codes to custom exceptions for clearer error reporting and management.
+    """
+
+    _exception_map = {}
```
Is this empty because the `AMDGPU` is yet to be fully implemented?

parthraut replied:
yes

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+        """Enables persistence mode for the specified GPU."""
+        raise ZeusNotSupportedGPUError(
+            "Persistence mode is not supported for AMD GPUs yet"
+        )
```
I don't know whether the notion of persistent mode exists for AMD GPUs, but if it doesn't exist and the driver is always just attached to the GPU, I think this should just be a no-op (i.e., `pass`).

parthraut replied:
I will look into it.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+    def getSupportedMemoryClocks(self) -> list[int]:
+        """Returns a list of supported memory clock frequencies for the specified GPU."""
+        num_supported, current, frequency = amdsmi.amdsmi_get_clk_freq(
+            self.handle, clk_type=amdsmi.AMDSMI_CLK_TYPE_SYS
```
```suggestion
            self.handle, clk_type=amdsmi.AmdSmiClkType.MEM
```
From https://rocm.docs.amd.com/projects/amdsmi/en/docs-5.7.1/py-interface_readme_link.html#amdsmi-get-clock-info
I won't look at `AMDGPU` any further.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+            self.handle,
+            minMemClockMHz,
+            maxMemClockMHz,
+            clk_type=amdsmi.AMDSMI_CLK_TYPE_SYS,
```
```suggestion
            clk_type=amdsmi.AmdSmiClkType.MEM,
```
There seem to be a lot of clock frequencies available (e.g., VGX, SYS, MEM). Please investigate how exactly they are different.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+
+    def _ensure_homogeneous(self) -> None:
+        """Ensures that all tracked GPUs are homogeneous in terms of name."""
+        if not self.gpus:
```
If you want to assume that `self.gpus` exist, it should be declared as an abstract property.

parthraut replied:
did you mean something like this?

```
class GPUs(abc.ABC):
@abc.abstractproperty
    def gpus(self) -> dict[int, GPU]:
        """Returns a dictionary of GPU objects being tracked."""
        pass
```
        
This doesn't seem to work for me, probably the syntax is wrong.

jaywonchung replied:
`abc.abstractproperty` is deprecated IIRC. Syntax seems correct except for the indent.

```python
@property
@abc.abstractmethod
def gpus(self) -> dict[int, GPU]:
    """Maps what to what?"""
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+        first_gpu_name = next(iter(self.gpus.values())).getName()
+        for gpu in self.gpus.values():
+            if gpu.getName() != first_gpu_name:
+                raise ZeusBaseGPUError("All GPUs must be of the same name.")
```
- Just convert to `set()` and check if len is larger than 1.
- When erroring out, show the user the names of GPUs detected.
- It shouldn't be base error.

```python
gpu_names = [gpu.getName() for gpu in self.gpus.values()]
# Both zero (no GPUs found) and one are fine.
if len(set(gpu_names)) > 1:
    raise Error(f"Heterogeneous GPUs found: {gpu_names}")
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+        """Sets the specified GPU as the current device for CUDA operations."""
+        torch.cuda.set_device(
+            index
+        )  # torch.cuda.set_device takes in re-indexed cuda index, not nvml index
```
Delete both. These are CUDA- and even PyTorch-level code that should not be in the driver wrapper.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+    """
+
+    @abc.abstractmethod
+    def __init__(self) -> None:
```
Signature doesn't match with `NVIDIAGPUs`.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+    GPU of CUDA index 0, use the index 0, and for CUDA index 2, use the index 1.
+
+    If gpus_to_track is specified, only the GPUs listed are set up. For continuation of the example above,
+    if `gpus_to_track=[1]`, only GPU 2 of CUDA index 2 is tracked. In this case, to access GPU of CUDA index 2, use index 1.
```
```suggestion
    if `gpus_to_track=[1]`, only GPU 2 of CUDA index 2 is tracked. In this case, to access GPU of CUDA index 2, use index 1.
    This is consistent with how `CUDA_VISIBLE_DEVICES` is conventionally handled.
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+            This list refers to the CUDA device indices as seen by the system. If not provided,
+            all available GPUs will be tracked. The indices in this list are affected by the
+            `CUDA_VISIBLE_DEVICES` environment variable, meaning that `gpus_to_track` should match
+            against the visible devices, not necessarily against the system-wide device indices.
```
```suggestion
        gpu_indices (list[int], optional): A list of CUDA indices specifying which GPUs to track. Respects `CUDA_VISIBLE_DEVICES`.
```
Changed parameter name to be consistent with `ZeusMonitor`. Also, you've already explained `CUDA_VISIBLE_DEVICES`, so made it shorter.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+            `CUDA_VISIBLE_DEVICES` environment variable, meaning that `gpus_to_track` should match
+            against the visible devices, not necessarily against the system-wide device indices.
+        ensure_homogeneous (bool, optional): Ensures that all GPUs being tracked are homogeneous (i.e.,
+            of the same model and with the same specifications).
```
```suggestion
        ensure_homogeneous (bool, optional): Ensures that all GPUs being tracked are homogeneous (i.e.,
            have the same output of `getName`).
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+
+    Attributes:
+        visible_indices (list[int]): The list of GPU indices that are visible and considered for tracking,
+            derived from `CUDA_VISIBLE_DEVICES` environment variable or system-wide visible devices if `CUDA_VISIBLE_DEVICES` is not set.
```
I actually don't understand what this means. The phrase *derived from* is confusing. I think the docstring should rather clearly define **NVML GPU indices** (which is in the order of physical GPUs attached) and **CUDA GPU indices** (which is a subset of GPUs influenced by `CUDA_VISIBLE_DEVICES`) and use these terminologies to indicate whether a certain integer index is NVML or CUDA.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+            for index, gpu_num in enumerate(self.visible_indices):
+                self.gpus[index] = NVIDIAGPU(gpu_num)
+
+    def __init__(
```
`__init__` should come first.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+        are visible based on the `CUDA_VISIBLE_DEVICES` environment variable and the optional `gpus_to_track`
+        parameter. It ensures that only the specified GPUs (if any) are tracked and managed by this instance.
+
+        Parameters:
```
Info is duplicated across the class docstring and method docstring. These should actually just be in `__init__`.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+class AMDGPUs(GPUs):
+    """Represents a collection of AMD GPUs for management and interaction, abstracting amdsmi calls and handling related exceptions.
+
+    This class provides a high-level interface to interact with AMD GPUs, taking into account the environment variable `ROCR_VISIBLE_DEVICES`
```
https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html#hip-visible-devices
There's also `HIP_VISIBLE_DEVICES`, which is interpreted the same as `CUDA_VISIBLE_DEVICES`. Please investigate.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+    If gpus_to_track is specified, only the GPUs listed are set up. For continuation of the example above,
+    if `gpus_to_track=[1]`, only GPU 2 of ROCR index 2 is tracked. In this case, to access GPU of ROCR index 2, use index 1.
+
+    Parameters:
```
Ditto comment regarding info duplication, conciseness, and clear distinction of system-level indices and application-level indices.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+        super().__init__(gpu_index)
+
+
+class AMDGPU(GPU):
```
I won't look too deeply into AMD GPUs for now. I guess we'll understand stuff better once we start running things there.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+            This is raised when initialization fails due to an NVML error, with specifics provided by the NVML exception.
+    """
+
+    def _init_gpus(self, gpus_to_track: list[int] = None) -> None:
```
```suggestion
    def _init_gpus(self, gpus_to_track: list[int] | None = None) -> None:
```
Second time spotting this mistake. Please review the codebase to fix all of these.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+                f"NVIDIA Error: {nvidia_error}\n"
+                f"AMD Error: {amd_error}"
+            ) from None
+    return _gpus
```
The core problem of this is that it's not scalable to adding more types of GPUs. It'll just keep nesting and nesting under `try-except` statements.

Please adopt the conventions in:
https://github.com/ml-energy/zeus/blob/f9275786fd98920b3bf98485c78f9e15b2304364/zeus/util/framework.py#L28-L42

Then, do `if-elif-else` on `nvml_is_available()` and `amdsmi_is_available()`.

jaywonchung commented on zeus/monitor/energy.py:
```diff
@@ -176,26 +168,21 @@ def __init__(
         else:
             self.power_monitor = None
 
+    # NOTE: right now, supports `getTotalEnergyConsumption` API is cached here. Should be cached in GPUs?
```
Oh yeah, I think moving `lru_cache` to `supportsGetTotalEnergyConsumption` makes sense. Then probably just remove stuff related to *new architecture*. It was more of a proxy to determine whether total energy consumption measurement is supported. Now that we have a method exactly for that, all these can be removed.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+        pass
+
+    @abc.abstractmethod
+    def getPowerUsage(self) -> int:
```
For every method, please mention the unit. For power, it's `mW`, for example.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+        gpus_to_track (Optional[list[int]]): A list of integers representing the GPU indices
+                                             to track. If None, all GPUs are tracked.
+        ensure_homogeneous (bool, optional): If True, attempts to ensure that all tracked GPUs are of the same model
+                and specifications.
```
```suggestion
        ensure_homogeneous (bool, optional): If True, raises an error if the names of all GPUs are not homogeneous.
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+            amdsmi.amdsmi_shut_down()  # Ignore error on shutdown. Neccessary for proper cleanup and test functionality
+
+
+_gpus = None
```
```suggestion
_gpus: GPUs | None = None
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,908 @@
+_gpus = None
+
+
+def get_gpus(gpus_to_track: list[int] = None, ensure_homogeneous: bool = True) -> GPUs:
```
What if someone calls
```python
get_gpus([1])
```
and then calls the following somewhere else?
```python
get_gpus([0, 2])
```

I think it'll return an instance that just holds GPU index 1. So this an error-prone API.

I wonder what the use case of being able to specify `gpu_indices` is. Why not just remove that option and always initialize all visible GPUs (constrained by `CUDA_VISIBLE_DEVICES`)?
The user anyway has to pass in `index` when calling GPU management APIs on `GPUs`. If they know the specific GPU indices they would like to use, they'll set `index` when calling methods. And the `index` of GPUs the user does not care about will just not be passed in. Then some GPU handles will never be used after being initialized, but it's negligible. WDYT?

parthraut replied:
Yes, that's how it was initally done. But when the unit tests ran, it looks like they only created mock handles for gpus that were listed in gpu_indices:

```
def mock_pynvml(nvml_indices: list[int], archs: list[int]) -> None:
        assert len(nvml_indices) == len(archs)
        index_to_handle = {i: f"handle{i}" for i in nvml_indices}
        handle_to_arch = {f"handle{i}": arch for i, arch in zip(nvml_indices, archs)}
        pynvml_mock.nvmlDeviceGetCount.return_value = NUM_GPUS
        pynvml_mock.nvmlDeviceGetHandleByIndex.side_effect = lambda index: index_to_handle[index]
        pynvml_mock.nvmlDeviceGetArchitecture.side_effect = lambda handle: handle_to_arch[handle]
```
        
So `index_to_handle` would repeatedly encounter a key error. I thought it was incorrect behavior to initialize a GPU that is never used, so I implemented it with gpu_indices. But this can be changed, but then the testing code needs to be modified to provide handles for all gpus.

jaywonchung replied:
Existing testing code should not stop us from implementing the correct API. Please fix the tests.

jaywonchung reviewed (commented):
Thanks again for your work! It's such a fundamentally large piece of work you did.

I mainly looked into `gpu.py`, and focused on NVIDIA GPUs, and left some comments and suggestions. After this round of comment is addressed, I'll also look at the places that use the new GPU device management API.

parthraut committed:
small changes from review

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update tests/test_monitor.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update zeus/device/gpu.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

jaywonchung commented:
Use the "Add to batch" feature

parthraut committed:
made a TODO

parthraut committed:
remove cuda_sync and setDevice from gpu, put it back where it belongs

parthraut committed:
changed gpu/get_gpu to torch_is_available()-like to make it scalable

parthraut committed:
✅ Fixed set Persistence Mode, added TODO comment back to where set Persistence mode SYS_admin check was
    ✅ included ability to turn off persistence mode

parthraut committed:
✅ Fixed set PowerManagementLimit - originally broke API consistency. Added resetPowerManagementLimit function.

parthraut committed:
✅ Fixed: To check ensure homogeneous, just convert to set and check len larger than 1

parthraut committed:
✅ Fixed names of exceptions

parthraut committed:
✅ Fall back to ZeusGPUUnknown Error

parthraut committed:
BIG BOY fix 1: ✅ move lru_cache stuff to gpu Class

parthraut committed:
init

jaywonchung commented:
Apparently [this](https://github.com/squidfunk/mkdocs-material/issues/6983) is what's causing doc build failures. Don't worry about doc failing and I think it'll be resolved automatically soon.

parthraut committed:
fixed tests

parthraut committed:
BIG BOY fix 2: ✅ Fix tests, and remove gpu_indicies

parthraut committed:
fixed style, added better doc strings

parthraut committed:
small style changes, ready for review

parthraut commented:
Ready for review!

Completed tasks:

✅ remove cuda_sync and setDevice from gpu, put it back where it belongs 
✅ changed gpu/get_gpu to torch_is_available()-like to make it scalable

✅ Fixed set Persistence Mode, added TODO comment back to where set Persistence mode SYS_admin check was
    ✅ included ability to turn off persistence mode

✅ Fixed set PowerManagementLimit - originally broke API consistency. Added resetPowerManagementLimit function. 
✅ Fixed: To check ensure homogeneous, just convert to set and check len larger than 1 
✅  declare self.gpus as an abstract property to ensure it exists

✅ Fixed names of exceptions
✅ Fall back to ZeusGPUUnknown Error

BIG fixes:
✅ move lru_cache stuff to gpu Class
✅ Fix tests, and remove gpu_indicies

STYLE Fixes:
✅ Module-level docstring in zeus/__init__.py and zeus/device/__init__.py
✅ Fixed docstrings
    ✅  Removed all private attributes from docstrings

✅ made sure signature of __init__ all match
✅ __init__ should come First
✅ mention unit for each method

Questions and notes:
❓ It looks like it is failing on github actions, but it works perfectly locally. I'm not sure what exactly is the issue.
❓ What do you think about the names of exceptions?

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,829 @@
+    }
+
+    @staticmethod
+    def _handle_nvml_errors(func):
```
Try removing `staticmethod` and moving this outside of the class. The test failure has to do something about this. Otherwise, I don't know. Either your local environment is not representative (e.g., like last time when you had PyTorch installed and tests passed locally) or testing is not good, you need to investigate.

parthraut replied:
that fixed it, all passing now

jaywonchung commented on TODO.txt:
```diff
@@ -0,0 +1,47 @@
+Completed:
```
Remove. And avoid doing `git add .`

parthraut replied:
oops, removed

jaywonchung reviewed (commented):
I'm outside so I just added two comments. I'll continue.

jaywonchung commented:
Exception names look good! But is there a need to export all of the exceptions in `zeus/devices/__init__.py`? I think just the core API like `get_gpu` should be exported and other things kept inside zeus.device.gpu.

parthraut committed:
moved handle outside of classes, removed TODO

parthraut commented:
> Exception names look good! But is there a need to export all of the exceptions in `zeus/devices/__init__.py`? I think just the core API like `get_gpu` should be exported and other things kept inside zeus.device.gpu.

So initially I was only exporting the exceptions used in the source code. But you said not to selectively export some of them, so I added all of them.

jaywonchung commented on zeus/__init__.py:
```diff
@@ -22,6 +22,7 @@
 - [`analyze`][zeus.analyze]: Functions for analyzing log files.
 - [`job`][zeus.job]: Job specification.
 - [`simulate`][zeus.simulate]: Machinery for trace-driven Zeus.
+- [`device`][zeus.device]: Abstraction of GPU devices.
```
```suggestion
- [`device`][zeus.device]: Abstraction of compute devices.
```

jaywonchung commented on tests/test_monitor.py:
```diff
@@ -177,6 +179,9 @@ def get_energy(self, start: float, end: float) -> dict[int, float]:
     log_file = tmp_path / "log.csv"
 
+    # want to make zeus.device.gpu.pynvml_is_available is a function, want it to always return true when testing
+    mocker.patch("zeus.device.gpu.pynvml_is_available", return_value=True)
```
```suggestion
    mocker.patch("zeus.device.gpu.nvml_is_available", return_value=True)
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,825 @@
+import pynvml  # necessary for testing to mock!
+
+from zeus.device.exception import ZeusBaseGPUError
+from zeus.util import pynvml_is_available, amdsmi_is_available
```
These functions should be in this file.

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,825 @@
+        pynvml.NVML_ERROR_NOT_FOUND: ZeusGPUNotFoundError,
+        pynvml.NVML_ERROR_INSUFFICIENT_SIZE: ZeusGPUInsufficientSizeError,
+        pynvml.NVML_ERROR_INSUFFICIENT_POWER: ZeusGPUInsufficientPowerError,
+        pynvml.NVML_ERROR_DRIVER_NOT_LOADED: ZeusGPUDriverError,  # change to : ZeusGPUDriverNotLoadedError
```
Noting TODO comment

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,825 @@
+        return _gpus
+    else:
+        raise ZeusGPUInitError(
+            "Failed to initialize GPU monitoring for NVIDIA and AMD GPUs."
```
```suggestion
            "NVML and AMDSMI unavailable. Failed to initialize GPU management library."
```

jaywonchung commented on zeus/monitor/power.py:
```diff
@@ -76,14 +74,16 @@ def infer_counter_update_period(nvml_handles: list[pynvml.c_nvmlDevice_t]) -> fl
+def _infer_counter_update_period_single(gpu_index: int) -> float:
     """Infer the update period of the NVML power counter for a single GPU."""
+    # get gpus
+    gpus = get_gpus()
```
Dang, this is so convenient. Nice job.

jaywonchung commented on zeus/monitor/power.py:
```diff
@@ -128,20 +128,24 @@ def __init__(
-        # Initialize NVML.
-        pynvml.nvmlInit()
+        # Get GPUs
+        gpus = get_gpus(gpu_indices)
```
`gpu_indices`? How did this get past Ruff?

jaywonchung replied:
```suggestion
        gpus = get_gpus()
```

jaywonchung commented on zeus/monitor/power.py:
```diff
@@ -128,20 +128,24 @@ def __init__(
+        # Get GPU indices:
+        # if gpu_indices is not None, we can directly use these indices in the gpu class,
+        # since the class only tracks all CUDA_VISIBLE_DEVICES gpus, or all gpus if CUDA_VISIBLE_DEVICES is not set
+        # if gpu_indices is None, we must track all GPUs visible under CUDA_VISIBLE_DEVICES (which the gpus class keeps track of)
```
```suggestion
        # Get GPUs
```
I would say the content of the comment is implicit when the user users the `GPUs` API.

jaywonchung commented on zeus/monitor/power.py:
```diff
@@ -239,27 +242,25 @@ def get_power(self, time: float | None = None) -> dict[int, float] | None:
 
 def _polling_process(
-    nvml_gpu_indices: list[int],
+    gpu_indices: list[int],
```
Just checking my understanding with you -- Since the `NVIDIAGPUs` class will take of the NVML-CUDA device index mapping, now everything can just forget about NVML indices and all operate on CUDA indices, right?

jaywonchung commented on zeus/optimizer/perseus/optimizer.py:
```diff
@@ -160,7 +152,7 @@ def __init__(
 
         # The frequency controller is responsible for controlling the frequency
         # of the GPU (nvml_device_id) asynchronously.
```
```suggestion
        # of the GPU (device_id) asynchronously.
```

jaywonchung commented on zeus/optimizer/power_limit.py:
```diff
@@ -271,13 +270,10 @@ def __init__(
         # Set the range of power limits to explore.
         # Assert that supported power limits ranges are uniform across GPUs.
-        pynvml.nvmlInit()
+        gpus = get_gpus()
```
```suggestion
        gpus = get_gpus(ensure_homogeneous=True)
```

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -0,0 +1,825 @@
+        """Instantiates NVIDIAGPUs object, setting up tracking for specified NVIDIA GPUs.
+
+        Args:
+            ensure_homogeneous (bool, optional): If True, ensures that all tracked GPUs have the same name (return value of pynvml.nvmlDeviceGetName). False by default.
```
```suggestion
            ensure_homogeneous (bool): If True, ensures that all tracked GPUs have the same name (return value of `nvmlDeviceGetName`). False by default.
```

jaywonchung commented on zeus/run/dataloader.py:
```diff
@@ -417,19 +416,16 @@ def __init__(
-        # Initialize NVML and get GPU handle or each GPU at the master process.
-        self.gpu_handles = []
-        pynvml.nvmlInit()
+        gpus = get_gpus()
```
```suggestion
        gpus = get_gpus(ensure_homogeneous=True)
```

jaywonchung commented on zeus/run/dataloader.py:
```diff
@@ -417,19 +416,16 @@ def __init__(
-        )
+        min_pl, self.max_pl = gpus.getPowerManagementLimitConstraints(
+            0
+        )  # TODO: assume 0?
```
```suggestion
        min_pl, self.max_pl = gpus.getPowerManagementLimitConstraints(0)
```

jaywonchung commented on zeus/run/master.py:
```diff
@@ -103,14 +103,12 @@ def __init__(
+        gpus = get_gpus()
+        minmax = gpus.getPowerManagementLimitConstraints(
+            0
+        )  # unit: mW, assumes device 0
```
```suggestion
        gpus = get_gpus(ensure_homogeneous=True)
        minmax = gpus.getPowerManagementLimitConstraints(0)  # unit: mW
```

jaywonchung commented on zeus/util/framework.py:
```diff
@@ -57,3 +57,36 @@ def cuda_sync(device: int | None = None) -> None:
+
+# Should MODULE_CACHE be used here?
+
+
```
```suggestion
```

jaywonchung reviewed (commented):
Just some small change suggestions. After they're all addressed (or accepted), I think we'd be good to merge without further changes. Thanks again for the great work. Now it's so much easier to interface with the GPU!!

parthraut committed:
Apply suggestions from code review

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
fixed error that was causing tests to fail

parthraut committed:
Apply suggestions from code review

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
moved _is_avail func to gpu.py

jaywonchung commented on zeus/device/gpu.py:
```diff
@@ -303,7 +305,7 @@ def __init__(self, gpu_index: int) -> None:
         pynvml.NVML_ERROR_INSUFFICIENT_SIZE: ZeusGPUInsufficientSizeError,
         pynvml.NVML_ERROR_INSUFFICIENT_POWER: ZeusGPUInsufficientPowerError,
-        pynvml.NVML_ERROR_DRIVER_NOT_LOADED: ZeusGPUDriverError,  # change to : ZeusGPUDriverNotLoadedError
+        pynvml.NVML_ERROR_DRIVER_NOT_LOADED: ZeusGPUDriverNotLoadedError,  # change to : ZeusGPUDriverNotLoadedError
```
```suggestion
        pynvml.NVML_ERROR_DRIVER_NOT_LOADED: ZeusGPUDriverNotLoadedError,
```

jaywonchung commented on zeus/util/__init__.py:
```diff
@@ -18,3 +18,4 @@
 from zeus.util.logging import FileAndConsole
 from zeus.util.lr_scaler import LinearScaler, SquareRootScaler
 from zeus.util.metric import zeus_cost, ZeusCostThresholdExceededError
+from zeus.util.framework import cuda_sync
```
```suggestion
```

jaywonchung committed:
Apply suggestions from code review

jaywonchung reviewed (approved):
Let's fix the doc error on master. LGTM, thanks!!

jaywonchung committed:
Update zeus/monitor/power.py
